### ISSUE TITLE
Impossible to upgrade NVDA when it is running on another user account.

### ISSUE THREAD

**Issue by lukaszgo1** [nvaccess/nvda#9178]
### Steps to reproduce:
1. You need to have ad least 2 user accounts on your computer.
2. Start NVDA on the first account.
3. Log in into the second one and either perform update or installation from the launcher.

### Actual behavior:
During install process NVDA installer complains, that the NVDA.exe file is in use and cannot be overwritten. If the install is cancelled screen reader is no longer in a usable state,
### Expected behavior:
One of the:
1. Before starting the install check, if NVDA is running on another account, warn the user, and do not allow installation.
2. Better  solution would be to kill the NVDA on the other account and continue install. NVDA has administrator permission during install, so this should be possible.
#### NVDA installed/portable/running from source:
Installed
#### NVDA version:
Any
#### Windows version:
Any
#### Name and version of other software in use when reproducing the issue:
N/A
#### Does the issue still occur after restarting your PC?
Yes
#### Have you tried any other versions of NVDA?
N/a

**Comment by leonardder** [comment]
[**@lukaszgo1**](https://github.com/lukaszgo1) commented on [21 Jan 2019, 19:23 CET](https://github.com/nvaccess/nvda/issues/9178#issue-401466054 "2019-01-21T18:23:00Z - Replied by Github Reply Comments"):
> 1. Before starting the install check, if NVDA is running on another account, warn the user, and do not allow installation.

I'd prefer this solution. Of course, this warning can be a question that allows you to shut down the running copies of NVDA. This *never* should be done automatically though.

**Comment by leonardder** [comment]
I just stumbled upon this and found it really irritating, as it ran my copy of NVDA unusable. Therefore I'll take this.

**Comment by leonardder** [comment]
I did some investigations:

1. One way we could do this is:

	* Enumerate all processes on the system
	* Find out what processes are our executable
	* Terminate these processes

	Of course, all ***after*** confirmation

2. I tried to find out what the taskkill command does when trying to terminate NVDA gracefully. It seem to cause some stress for the NVDA process, as after executing the command multiple times, I get a fatal error bang and a major freeze of the system that requires me to do a sign out / sign in. This is definitely something to improve on.

**Comment by leonardder** [comment]
Another update. We can easily get the list of processes using WMI. We can even terminate them using WMI while at it, but that's something I do not prefer to do, as it adds unneeded complexity.

**Comment by feerrenrut** [comment]
Another approach might be to ensure that NVDA exits when the user logs out or switches.

**Comment by lukaszgo1** [comment]
@feerrenrut wrote:
> 
> Another approach might be to ensure that NVDA exits when the user logs out or switches.



This is IMHO very poor workaround. When logging off NVDA turns of as all other processes so in this case problem obviously does not exists. When switching users however you would expect all opened programs to stay intact when you've switched accounts and I do not  see a reason to make an exception for NVDA.

**Comment by leonardder** [comment]
at the logon screen turned on. After that, you try to update NVDA using an RDP session. That's not going to work as the logon screen of the system itself is still active with NVDA running.

I have a prototype in such a state that it can list processes for other users, it only fails to terminate them. Interesting enough, termination succeeds if I run nvda_slave elevated, so not sure what's going wrong here.

**Comment by leonardder** [comment]
Another update.

It seems that either there's something wrong with how we elevate, or something else.

1. If I run an elevated nvda_slave that tries to terminate a process for another user, I get access denied when trying to open a handle to the process. If I start nvda_slave from an elevated command prompt, it is able to open the handle and even terminate the process.
2. Same applies when I use WMI to terminate the process
3. However, when I call the taskkill process from an elevated nvda_slave, all is well and I'm able to terminate just fine.

**Comment by feerrenrut** [comment]
@lukaszgo1 Of course, this has to be transparent to the user. NVDA would need to be able to restore itself on user switching back. It would be good if this took into account the state of NVDA. A user may have modified but unsaved preferences, we can't assume they should be saved or discarded. Searching for and force killing the process doesn't seem like a good solution either. 

Handling the log-on screen is an interesting point, and I expect a much more likely problem to encounter than multiple users. I wonder if we could use a helper process that detects activity and starts NVDA then exits when the sign on screen disappears. I think we can detect the difference between these two screens.

Finally, it would be worth looking into how other applications handle this. My guess is that they don't, and it's expected that you close all instances of the application before updating. I use multiple profiles with Firefox, when it updates only one of the profiles keeps working until I restart all instances.

Another option again is to allow NVDA to install to a staging directory which can be used until the system is restarted. Then during restart the old is removed and the staged version replaces it. This will also help us with a silent auto update feature.

**Comment by lukaszgo1** [comment]
@feerrenrut  wrote:

> Searching for and force killing the process doesn't seem like a good solution either.

As long as user is asked if he/she wants to close other instances of NVDA I believe that it is the best solution. It is certainly good first step - in the current state if NVDA is running on another account and upgrade is attempted it is broken so hardly ideal. 

> Finally, it would be worth looking into how other applications handle this. My guess is that they don't, and it's expected that you close all instances of the application before updating. I use multiple profiles with Firefox, when it updates only one of the profiles keeps working until I restart all instances.



I cannot recall the specific software however I've seen multiple times a message stating that the program is running and question for the user if he is happy with it being closed. Again the simplest solution and fixing this in all use cases.

**Comment by feerrenrut** [comment]
>  I've seen multiple times a message stating that the program is running and question for the user if he is happy with it being closed

This is usually when the application is running for the current user.

> Again the simplest solution and fixing this in all use cases.

The process running as another user significantly complicates this, it seems @leonardder is running into some difficulties with this approach. It's worth considering other approaches.

**Comment by leonardder** [comment]
I'm not very comfortable with auto force killing either, so that's why I made it optional in my prototype.

I think I'd like to go with the easiest solution, namely checking whether there are any running processes and if so, refuse to install. We should also be able to give a hint about what logged in user is using the proces.

**Comment by feerrenrut** [comment]
@leonardder That sounds good, and keeps NVDA usable in this situation. Eventually we plan to streamline the update process, ideally in the background without disturbing the user (if enabled in config). In order to do that, we will likely need to be able to install to another location which takes over once NVDA is closed. For this we will also need to know if NVDA is running on other profiles.
